labelbox/schema/role.py

```diff
@@ -0,0 +1,69 @@
+from dataclasses import dataclass
+
+from labelbox.orm.model import Field
+from labelbox.orm.db_object import DbObject
+from labelbox.schema.project import Project
+
+
+class Roles:
+    """
+    Object that manages org and user roles
+
+        >>> roles = client.get_roles()
+        >>> roles.valid_roles # lists all valid roles
+        >>> roles['ADMIN'] # returns the admin Role
+
+    """
+    _instance = None
+
+    def __new__(cls, client):
+        if cls._instance is None:
+            cls._instance = super(Roles, cls).__new__(cls)
+            query_str = """query GetAvailableUserRolesPyApi { roles { id name } }"""
+            res = client.execute(query_str)
+            valid_roles = set()
+            for result in res['roles']:
+                _name = result['name'].upper().replace(' ', '_')
+                result['name'] = _name
+                setattr(cls._instance, _name, Role(client, result))
+                valid_roles.add(_name)
+            cls._instance.valid_roles = valid_roles
+        cls._instance.index = 0
+        return cls._instance
```

Comment:
In general I'm not a huge fan of overriding `__new__`. Imho the code gets overly complicated. If we want to make them singletons, could we instead declare the roles as lazy initalized module vars - sth like here https://github.com/Labelbox/python-monorepo/blob/6c6bfebb68fc6f876c3abf29c9ce727a6c5744dc/services/prediction_import/prediction_import/db/client.py#L357

Reply:
This is a pretty standard singleton pattern. The code snippet you provided is much simpler but also because it is doing much less. This is going to basically look the same except swap __new__ for __init__ and add a new module level variable which feels like a lateral move rather than an improvement.

Reply:
One thing I would say here, is it might be easier to grok this code if you moved the business logic into a named function
```python
def get_available_roles():
 ...
```

Reply:
I lean more towards, structs/class should be dumb containers of data than locations of business logic.

In that sense I think you could seperate fetching `GetAvailableUserRolesPyApi` logic from the class `__new__` and move it out into a static method or function, imho it doesn make reading the class a little trickier.

```python
class RoleAccessor(self):
    def __init__(self):
         self.roles = get_The_roles(client)
    
```

or some other means to break up the logic

Reply:
I changed it to this for now. https://github.com/Labelbox/labelbox-python/blob/e6e4dd6d759e4c4dea61f21735863a1a6cbe4bd3/labelbox/schema/role.py#L8-L52. I'm not totally clear what Nasr is suggesting so we are going to sync tomorrow.